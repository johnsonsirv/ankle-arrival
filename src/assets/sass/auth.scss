.loginPanel {
    display: flex;
    width: 300px;
    height: 280px;
    flex-direction: column;
    justify-content: space-evenly;
    border: 1px solid #d8d4d4;
    border-radius: 5px;
    margin: auto;
    margin-top: 10%;
    background-color: #ffffffc4;
    & .loginForm {
        display: flex;
        flex-direction: column;
        justify-content: space-evenly;
        height: 180px;
        & button {
            width: 12vw;
            height: 7vh;
            cursor: pointer;
        }
    }
    label {
        font-size: 13px;
        letter-spacing: .15em;
    }
}

.signUpPanel {
    display: flex;
    width: 300px;
    height: 370px;
    flex-direction: column;
    justify-content: space-evenly;
    border: 1px solid #d8d4d4;
    border-radius: 5px;
    margin: auto;
    margin-top: 10%;
    background-color: #ffffffc4;
    & .signUpForm {
        display: flex;
        flex-direction: column;
        justify-content: space-evenly;
        height: 290px;
        & button {
            margin-top: 3px;
            width: 12vw;
            height: 7vh;
            cursor: pointer;
        }
    }
    label {
        font-size: 12px;
        letter-spacing: .15em;
    }
}