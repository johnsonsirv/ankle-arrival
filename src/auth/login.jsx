/* eslint-disable jsx-a11y/label-has-associated-control */
import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Redirect } from 'react-router-dom';
import Spinner from 'react-spinkit';
import Joi from 'joi-browser';
import * as dispatchActions from '../actions';
import InputTextField from '../components/forms/inputs/inputTextField';
import Button from '../components/forms/inputs/button';
import SocialLoginPanel from './socialLoginPanel';
import {
  validateProperty,
  validateAllProperty,
  inputTextErrorStyle,
} from '../utils/validation';

const mapStateToProps = state => state;

export const Login = props => {
  const [account, setAccount] = useState({
    username: '',
    password: '',
    isValid: false,
    error: {
      username: null,
      password: null,
    },
  });

  const schema = {
    username: Joi.string().required().label('Username'),
    password: Joi.string().required().label('Password'),
  };

  const handleChange = e => {
    const { name, value } = e.target;
    account[name] = value;

    const data = { username: account.username, password: account.password };

    const error = validateProperty({ target: e.target, schema });
    account.isValid = validateAllProperty({ data, schema });

    if (error) {
      const errorProperty = error.path[0];
      account.error[errorProperty] = error.message;
    } else {
      account.error = {
        username: null,
        password: null,
      };
    }

    setAccount({ ...account });
  };
  const handleLogin = e => {
    e.preventDefault();
    const { username, password } = account;
    props.authenticateUser({
      username,
      password,
    });
  };

  const { getCurrentUser } = props;
  useEffect(() => {
    getCurrentUser();
  }, [getCurrentUser]);

  const {
    currentUser: { isAuthenticated, userLogin },
  } = props;
  const { error } = account;

  return (
    <>
      {isAuthenticated && <Redirect to="/doctors" />}
      {userLogin && userLogin.isFetching && (
        <Spinner name="three-bounce" fadeIn="none" />
      )}
      {userLogin && userLogin.invalid && <h4>Wrong Username or Password</h4>}
      <div className="loginPanel">
        <div>
          <SocialLoginPanel />
        </div>
        <div>
          <form className="loginForm" autoComplete="off">
            <div>
              <label htmlFor="username">Username</label>
              <br />
              <InputTextField
                type="text"
                id="username"
                name="username"
                onChange={handleChange}
                value={account.username}
                style={error && error.username ? inputTextErrorStyle : {}}
              />
            </div>
            <div>
              <label htmlFor="password">Password</label>
              <br />
              <InputTextField
                type="password"
                id="password"
                name="password"
                onChange={handleChange}
                value={account.password}
                style={error && error.password ? inputTextErrorStyle : {}}
              />
            </div>
            <div>
              <Button
                onClick={handleLogin}
                value="Login"
                disabled={!account.isValid}
                id="login"
              />
            </div>
          </form>
        </div>
      </div>
    </>
  );
};

Login.defaultProps = {
  currentUser: {},
};

Login.propTypes = {
  authenticateUser: PropTypes.func.isRequired,
  getCurrentUser: PropTypes.func.isRequired,
  currentUser: PropTypes.shape({
    id: PropTypes.number,
    username: PropTypes.string,
    token: PropTypes.string,
    isAuthenticated: PropTypes.bool,
    userLogin: PropTypes.shape({
      invalid: PropTypes.bool,
      isFetching: PropTypes.bool,
    }),
    userAccount: PropTypes.shape({
      created: PropTypes.bool,
    }),
  }),
};

export default connect(mapStateToProps, dispatchActions)(Login);
