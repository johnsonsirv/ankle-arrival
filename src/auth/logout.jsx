import React from 'react';

const Logout = () => {
  localStorage.removeItem('currentUser');
  window.location = '/';
  return <></>;
};

export default Logout;
