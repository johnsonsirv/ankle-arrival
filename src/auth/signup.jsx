import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Redirect } from 'react-router-dom';
import Spinner from 'react-spinkit';
import Joi from 'joi-browser';
import { toast } from 'react-toastify';
import * as dispatchActions from '../actions';
import InputTextField from '../components/forms/inputs/inputTextField';
import EmailTextField from '../components/forms/inputs/emailTextField';
import PasswordTextField from '../components/forms/inputs/passwordTextField';
import Button from '../components/forms/inputs/button';

const mapStateToProps = state => state;

export const Signup = props => {
  const [account, setAccount] = useState({
    firstname: null,
    lastname: null,
    email: null,
    username: null,
    password: null,
    city: 'Remote',
    isValid: false,
  });

  const schema = {
    firstname: Joi.string().required().label('Firstname'),
    lastname: Joi.string().required().label('Lastname'),
    email: Joi.string().email().required().label('Email'),
    username: Joi.string().required().label('Username'),
    password: Joi.string().required().label('Password'),
  };

  const validateProperty = ({ name, value }) => {
    const obj = { [name]: value };
    const subSchema = { [name]: schema[name] };
    const { error } = Joi.validate(obj, subSchema);
    return error ? error.details[0].message : null;
  };

  const handleChange = e => {
    const { name, value } = e.target;
    account[name] = value;
    const error = validateProperty(e.target);
    account.isValid = !error;

    setAccount({ ...account });
    toast.error(error);
  };
  const handleClick = e => {
    e.preventDefault();
    const { firstname, lastname, username, password, email, city } = account;
    props.createUserAccount({
      firstname,
      lastname,
      username,
      password,
      email,
      city,
    });
  };

  useEffect(() => {
    props.getCurrentUser();
  });

  const { currentUser: { userAccount, isAuthenticated } } = props;
  return (
    // render social login section here
    <>
      {isAuthenticated && <Redirect to="/doctors" />}
      {!userAccount.created && <Spinner name="three-bounce" fadeIn="none" />}
      <div>
        <form>
          <InputTextField
            id="firstname"
            name="firstname"
            onChange={handleChange}
            value={account.firstname}
          />
          <InputTextField
            id="lastname"
            name="lastname"
            onChange={handleChange}
            value={account.lastname}
          />
          <InputTextField
            id="username"
            name="username"
            onChange={handleChange}
            value={account.username}
          />
          <EmailTextField
            name="email"
            onChange={handleChange}
            value={account.email}
          />
          <PasswordTextField
            name="password"
            onChange={handleChange}
            value={account.password}
          />
          <Button
            onClick={handleClick}
            value="Sign Up"
            disabled={!account.isValid}
            id="signup"
          />
        </form>
      </div>
    </>
  );
};

Signup.defaultProps = {
  currentUser: {}
};

Signup.propTypes = {
  createUserAccount: PropTypes.func.isRequired,
  getCurrentUser: PropTypes.func.isRequired,
  currentUser: PropTypes.shape({
    id: PropTypes.number,
    username: PropTypes.string,
    token: PropTypes.string,
    isAuthenticated: PropTypes.bool,
    userLogin: PropTypes.shape({
      ok: PropTypes.bool,
    }),
    userAccount: PropTypes.shape({
      created: PropTypes.bool,
    })
  }),
};

export default connect(mapStateToProps, dispatchActions)(Signup);
