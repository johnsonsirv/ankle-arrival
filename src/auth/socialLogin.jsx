import React from 'react';

const SocialLogin = () => {
  return <></>;
};

export default SocialLogin;
