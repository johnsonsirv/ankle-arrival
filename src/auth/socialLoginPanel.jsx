import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { google, facebook, instagram } from '../config/socialLogin';
import * as dispatchActions from '../actions';
import SocialButton from '../components/forms/inputs/socialButton';

export const SocialLoginPanel = props => {
  const [user, setUser] = useState({});

  const handleSocialLogin = oauthUser => {
    const { firstname, email, lastname } = oauthUser;

    const userParams = {
      firstname,
      lastname,
      username: 'uniq_gen',
      email,
      city: 'Remote',
    };
    setUser(userParams);
    props.userFromOauth(user);
    console.log(user);
  };

  const handleSocialLoginFailure = err => {
    console.log(err);
  };

  return (
    <>
      <SocialButton
        provider={google.providerName}
        appId={google.appId}
        onLoginSuccess={handleSocialLogin}
        onLoinFailure={handleSocialLoginFailure}
        id="google-login"
        className=""
        value="Login with Google"
      />
      <SocialButton
        provider={facebook.providerName}
        appId={facebook.appId}
        onLoginSuccess={handleSocialLogin}
        onLoinFailure={handleSocialLoginFailure}
        id="facebook-login"
        className=""
        value="Login with Facebook"
      />
      {/* <SocialButton
        provider={twitter.providerName}
        appId={twitter.appId}
        onLoginSuccess={handleSocialLogin}
        onLoinFailure={handleSocialLoginFailure}
        id="twitter-login"
        className=""
        value="Login with Twitter"
      /> */}
      <SocialButton
        provider={instagram.providerName}
        appId={instagram.appId}
        onLoginSuccess={handleSocialLogin}
        onLoinFailure={handleSocialLoginFailure}
        id="instagram-login"
        className=""
        value="Login with Instagram"
      />
    </>
  );
};

SocialLoginPanel.defaultProps = {
  userFromOauth: null,
};

SocialLoginPanel.propTypes = {
  userFromOauth: PropTypes.func,
};

export default connect(null, dispatchActions)(SocialLoginPanel);
