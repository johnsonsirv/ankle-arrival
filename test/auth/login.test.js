/* eslint-disable react/jsx-props-no-spreading */
/* eslint-disable react/react-in-jsx-scope */
/* eslint-disable no-undef */
import { BrowserRouter as Router } from 'react-router-dom';
import { Login } from '../../src/auth/login';

function setup() {
  const props = {
    authenticateUser: jest.fn(),
    getCurrentUser: jest.fn(),
    isAuthenticated: false,
    userLogin: { ok: false },
  };

  const enzymeWrapper = shallow(<Login {...props} />);

  return { props, enzymeWrapper };
}

describe('<Login /> rendering', () => {
  const { enzymeWrapper } = setup();

  it('should render correctly', () => {
    expect(enzymeWrapper).toMatchSnapshot();
  });
  it('should render <SocialLoginPanel>', () => {
    expect(enzymeWrapper.find('SocialLoginPanel')).toHaveLength(1);
  });
  it('should render 1 <PasswordFieldText>', () => {
    expect(enzymeWrapper.find('PasswordTextField')).toHaveLength(1);
  });
  it('should render 1 signup <Button> ', () => {
    expect(enzymeWrapper.find('#login')).toHaveLength(1);
  });
  it('should render 1 <InputTextField /> for username', () => {
    expect(enzymeWrapper.find('#username')).toHaveLength(1);
  });
});

describe('<Login /> interactions', () => {
  const { props, enzymeWrapper } = setup();
  it('should enable <button> if inputs are valid', () => {
    enzymeWrapper.find('#username').simulate('change', {
      target: { value: 'anotheruser', name: 'username' },
    });

    enzymeWrapper.find('PasswordTextField').simulate('change', {
      target: { value: '1234', name: 'password' },
    });
    expect(enzymeWrapper.find('#login').prop('disabled')).toBe(false);
  });
  it('should disable <button> for click if inputs are invalid', () => {
    enzymeWrapper.find('#username').simulate('change', {
      target: { value: null, name: 'username' },
    });

    enzymeWrapper.find('PasswordTextField').simulate('change', {
      target: { value: null, name: 'password' },
    });
    expect(enzymeWrapper.find('#login').prop('disabled')).toBe(true);
  });

  it('<InputTextField /> should show username onChange', () => {
    enzymeWrapper.find('#username').simulate('change', {
      target: { value: 'anotheruser', name: 'username' },
    });
    expect(enzymeWrapper.find('#username').get(0).props.value).toMatch(
      /anotheruser/
    );
  });
  it('<PasswordTextField /> should show password input as *** onChange', () => {
    enzymeWrapper.find('PasswordTextField').simulate('change', {
      target: { value: '1234', name: 'password' },
    });
    expect(enzymeWrapper.find('PasswordTextField').get(0).props.value).toMatch(
      /1234/
    );
  });
  it('should dispatch authenticateUser action on handleLogin', () => {
    enzymeWrapper.find('#username').simulate('change', {
      target: { value: 'anotheruser', name: 'username' },
    });
    enzymeWrapper.find('PasswordTextField').simulate('change', {
      target: { value: '1234', name: 'password' },
    });
    enzymeWrapper
      .find('#login')
      .simulate('click', { target: {}, preventDefault: jest.fn() });
    expect(props.authenticateUser).toHaveBeenCalled();
  });
});

describe('<Login /> routing', () => {
  it('should redirect to doctors if user is already logged in', () => {
    const props = {
      authenticateUser: jest.fn(),
      getCurrentUser: jest.fn(),
      isAuthenticated: true,
      userLogin: { ok: false },
    };
    const enzymeWrapper = shallow(
      <Router>
        <Login {...props} />
      </Router>
    );
    expect(enzymeWrapper.find('doctors')).not.toHaveLength(1);
  });
});
