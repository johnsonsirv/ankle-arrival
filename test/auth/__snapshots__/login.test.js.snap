// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Login /> rendering should render correctly 1`] = `ShallowWrapper {}`;
